internal/testrunner/test_case_parser.go

```diff
@@ -42,19 +41,88 @@ var optionRegex = regexp.MustCompile(`(?m)^\/{2}\s*@(\w+)\s*:\s*([^\r\n]*)`)
 // Regex for parsing @link option
 var linkRegex = regexp.MustCompile(`(?m)^\/{2}\s*@link\s*:\s*([^\r\n]*)\s*->\s*([^\r\n]*)`)
 
+// File-specific directives used by fourslash tests
+var fourslashDirectives = []string{"emitthisfile"}
+
 // Given a test file containing // @FileName directives,
 // return an array of named units of code to be added to an existing compiler instance.
 func makeUnitsFromTest(code string, fileName string) testCaseContent {
+	testUnits, symlinks, currentDirectory, _ := ParseTestFilesAndSymlinks(
+		code,
+		fileName,
+		func(filename string, content string, fileOptions map[string]string) *testUnit {
+			return &testUnit{content: content, name: filename}
+		},
+	)
+	if currentDirectory == "" {
+		currentDirectory = srcFolder
+	}
+
+	// unit tests always list files explicitly
+	allFiles := make(map[string]string)
+	for _, data := range testUnits {
+		allFiles[tspath.GetNormalizedAbsolutePath(data.name, currentDirectory)] = data.content
+	}
+	parseConfigHost := tsoptionstest.NewVFSParseConfigHost(allFiles, currentDirectory, true /*useCaseSensitiveFileNames*/)
+
+	// check if project has tsconfig.json in the list of files
+	var tsConfig *tsoptions.ParsedCommandLine
+	var tsConfigFileUnitData *testUnit
+	for i, data := range testUnits {
+		if harnessutil.GetConfigNameFromFileName(data.name) != "" {
+			configFileName := tspath.GetNormalizedAbsolutePath(data.name, currentDirectory)
+			path := tspath.ToPath(data.name, parseConfigHost.GetCurrentDirectory(), parseConfigHost.Vfs.UseCaseSensitiveFileNames())
+			configJson := parser.ParseJSONText(configFileName, path, data.content)
+			tsConfigSourceFile := &tsoptions.TsConfigSourceFile{
+				SourceFile: configJson,
+			}
+			configDir := tspath.GetDirectoryPath(configFileName)
+			tsConfig = tsoptions.ParseJsonSourceFileConfigFileContent(
+				tsConfigSourceFile,
+				parseConfigHost,
+				configDir,
+				nil, /*existingOptions*/
+				configFileName,
+				nil, /*resolutionStack*/
+				nil, /*extraFileExtensions*/
+				nil /*extendedConfigCache*/)
+			tsConfigFileUnitData = data
+
+			// delete tsconfig file entry from the list
+			testUnits = slices.Delete(testUnits, i, i+1)
+			break
+		}
+	}
+
+	return testCaseContent{
+		testUnitData:         testUnits,
+		tsConfig:             tsConfig,
+		tsConfigFileUnitData: tsConfigFileUnitData,
+		symlinks:             symlinks,
+	}
+}
+
+// !!! move this to a different package
+// Given a test file containing // @FileName and // @symlink directives,
+// return an array of named units of code to be added to an existing compiler instance,
+// along with a map of symlinks and the current directory.
+func ParseTestFilesAndSymlinks[T any](
```

Comment:
I extracted this function from the old `makeUnitsFromTest` so that I can reuse it for fourslash test parsing.